### PR TITLE
FEATURE: LX-6 Multiple date filters in Dashboard

### PR DESCRIPTION
**Pull request by ivan-nejezchleb**
---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17171/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18667/ : FAILURE in 8m 23s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17179/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18676/ : FAILURE in 9m 30s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17189/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18684/ : FAILURE in 9m 05s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17190/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18685/ : FAILURE in 8m 58s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17216/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18721/ : FAILURE in 1m 50s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17217/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18722/ : FAILURE in 9m 04s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17277/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18763/ : FAILURE in 7m 21s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17284/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18767/ : FAILURE in 7m 24s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17286/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18769/ : FAILURE in 7m 19s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17300/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18784/ : FAILURE in 9m 59s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17304/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18788/ : FAILURE in 8m 20s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17316/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18794/ : FAILURE in 13m 17s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17332/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18814/ : FAILURE in 4m 25s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17333/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18815/ : FAILURE in 12m 20s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17335/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18817/ : SUCCESS in 14m 35s

**Comment by ivan-nejezchleb** [comment]
extended test - backstop

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2579/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17336/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1807/ : SUCCESS in 25m 55s

File: libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableFilterDropZone/DraggableFilterDropZone.tsx [hunk 1]
```diff
@@ -72,7 +84,7 @@ export function AttributeFilterDropZone({ targetIndex, onDrop }: AttributeFilter
                         </div>
                     </div>
                     <Bubble alignPoints={[{ align: "bc tc", offset: { x: 0, y: 0 } }]}>
-                        <FormattedMessage id="filterBar.filter.dropzone.tooltip" />
+                        <FormattedMessage id={tooltipId} />
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableFilterDropZone/DraggableFilterDropZone.tsx:87]: Isn't this the example that could use `defineMessage`?

File: libs/sdk-ui-dashboard/src/presentation/filterBar/attributeFilter/addAttributeFilter/AttributesDropdown.tsx [hunk 2]
```diff
@@ -125,10 +135,39 @@ export function AttributesDropdown({
...
         : attributes;
 
+    const offerDateFilters = enableMultipleDateFilters && supportsMultipleDateFilters;
+
+    const filteredDateDatasets = searchQuery
+        ? dateDatasets.filter((ds) => ds.dataSet.title.toLowerCase().includes(searchQuery.toLowerCase()))
+        : dateDatasets;
+
+    const [selectedTabId, setSelectedTabId] = useState(
+        attributes.length || !offerDateFilters ? "attributes" : "dateDatasets",
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/presentation/filterBar/attributeFilter/addAttributeFilter/AttributesDropdown.tsx:149]: Super minor, but you don't need negation if you switch the branches ;)

File: libs/sdk-ui-dashboard/styles/scss/attributeFilterConfig.scss [hunk 3]
```diff
@@ -25,61 +26,61 @@
...
-            position: absolute;
-            top: 50%;
-            display: inline-block;
-            width: 100%;
-            height: 0;
-            margin-left: 10px;
-            border-top: 1px solid var(--gd-palette-complementary-3, #dde4eb);
-        }
-    }
+    // .configuration-panel-header {
```
Comment (xMort) [inline libs/sdk-ui-dashboard/styles/scss/attributeFilterConfig.scss:29]: Please delete it or put there some TODO for later.

File: libs/sdk-ui-dashboard/src/model/queryServices/queryWidgetFilters.ts [hunk 4]
```diff
@@ -166,25 +167,108 @@ export function isDashboardDateFilterIgnoredForInsight(insight: IInsightDefiniti
...
+function resolveDateFilters(
+    insightDateFilterDateDatasetPairs: IFilterDateDatasetPair[],
+    dashboardDateFilterDateDatasetPairs: IFilterDateDatasetPair[],
+): IDateFilter[] {
+    // prioritize dashboard filters over insight ones
     // and strip useless all time filters at the end
-    return allDateFilterDateDatasetPairs
+    const init = dashboardDateFilterDateDatasetPairs
+        .filter((item) => !!item.dateDataset)
+        .reduce((acc: IDateFilter[], curr) => {
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/model/queryServices/queryWidgetFilters.ts:267]: fixed

File: libs/sdk-ui-dashboard/src/_staging/dashboard/dashboardFilterConverter.ts [hunk 5]
```diff
@@ -63,25 +64,33 @@ export function attributeFilterToDashboardAttributeFilter(
...
         tempDateDatasetId,
         excludeCurrentPeriod,
     );
 
     if (!afmFilter) {
-        return undefined;
+        // All time filter representation
+        return {
+            dateFilter: {
+                type: "relative",
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/_staging/dashboard/dashboardFilterConverter.ts:89]: AllTime definition and detection will be solved in separate PR

File: libs/sdk-ui-dashboard/src/model/store/filterContext/filterContextReducers.ts [hunk 6]
```diff
@@ -141,18 +146,35 @@ export type IUpsertDateFilterPayload = IUpsertDateFilterAllTimePayload | IUpsert
...
-     */
     if (action.payload.type === "allTime") {
         if (existingFilterIndex >= 0) {
-            // if allTime remove the date filter altogether
-            state.filterContextDefinition.filters.splice(existingFilterIndex, 1);
+            if (dateDataSet) {
+                const dateFilter = state.filterContextDefinition.filters[existingFilterIndex];
+
+                if (isDashboardDateFilter(dateFilter)) {
+                    dateFilter.dateFilter.type = "relative";
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/model/store/filterContext/filterContextReducers.ts:169]: AllTime definition and detection will be solved in separate PR

File: libs/sdk-ui-dashboard/src/_staging/dateFilterConfig/dateFilterOptionMapping.ts [hunk 7]
```diff
@@ -51,17 +51,25 @@ export function matchDateFilterToDateFilterOptionWithPreference(
         ? findDateFilterOptionById(preferredOptionId, availableOptions)
         : undefined;
 
+    const isAllTime =
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/_staging/dateFilterConfig/dateFilterOptionMapping.ts:54]: AllTime definition and detection will be solved in separate PR

File: libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableFilterDropZone/DraggableFilterDropZone.tsx (hunk 1 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/presentation/dragAndDrop/draggableFilterDropZone/DraggableFilterDropZone.tsx:87]: fixed

File: libs/sdk-ui-dashboard/src/presentation/filterBar/attributeFilter/addAttributeFilter/AttributesDropdown.tsx (hunk 2 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/presentation/filterBar/attributeFilter/addAttributeFilter/AttributesDropdown.tsx:149]: fixed

File: libs/sdk-ui-dashboard/styles/scss/attributeFilterConfig.scss (hunk 3 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/styles/scss/attributeFilterConfig.scss:29]: removed as replaced by shared styles

File: libs/sdk-ui-dashboard/src/model/store/filterContext/filterContextReducers.ts [hunk 8]
```diff
@@ -486,6 +509,86 @@ const changeSelectionMode: FilterContextReducer<PayloadAction<IChangeAttributeSe
...
+const addDateFilter: FilterContextReducer<PayloadAction<AddDateFilterPayload>> = (state, action) => {
+    invariant(state.filterContextDefinition, "Attempt to edit uninitialized filter context");
+
+    const { index, dateDataset } = action.payload;
+
+    const filter: IDashboardDateFilter = {
+        dateFilter: {
+            dataSet: dateDataset,
+            type: "relative",
+            granularity: "GDC.time.date",
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/model/store/filterContext/filterContextReducers.ts:521]: yes, AllTime definition and detection will be solved in separate PR

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-ui-dashboard/src/model/store/dateFilterConfig/const.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// (C) 2024 GoodData Corporation
+
+export const DEFAULT_DASHBOARD_DATE_FILTER_NAME = "Date range";
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply libs/sdk-ui-dashboard/src/model/store/dateFilterConfig/const.ts]: fixed

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17345/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18833/ : FAILURE in 13m 14s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17346/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18834/ : FAILURE in 2m 07s

**Review by xMort — APPROVED** [review]

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by xMort** [comment]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR4448

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3639/ : SUCCESS in 8m 58s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18836/ : SUCCESS in 12m 02s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1264/ : SUCCESS in 11m 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4804/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/1518/ : FAILURE in 1m 19s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1206/ : SUCCESS in 14m 49s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/562/ : SUCCESS in 6m 20s

**Pull request merged**